Compile error with gcc 8

## Required information

**Operating system:**
Ubuntu 20.04 LTS

**Compiler version:**
GCC 8.4.0

**Eclipse iceoryx version:**
`main` branch

**Observed result or behaviour:**
A clear and precise description of the observed result.
```
iceoryx_hoofs/reporting/include/iox/assertions.hpp:83:26: error: call to non-'constexpr' function 'void iox::er::forwardPanic(const iox::er::SourceLocation&, Message&&) [with Message = const char (&)[50]]'
     iox::er::forwardPanic(IOX_CURRENT_SOURCE_LOCATION, "Reached code that was supposed to be unreachable.")
     ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     note: in expansion of macro 'IOX_UNREACHABLE'
     IOX_UNREACHABLE();
     ^~~~~~~~~~~~~~~
```


**Expected result or behaviour:**
no compile error

**Conditions where it occurred / Performed steps:**
Compiling with gcc 8, e.g. for CUDA

## Additional helpful information

see also https://github.com/eclipse-iceoryx/iceoryx2/issues/855

--- a/doc/website/release-notes/iceoryx-unreleased.md
+++ b/doc/website/release-notes/iceoryx-unreleased.md
@@ -159,6 +159,7 @@
 - Max Client and Server cannot be configured independently of Max Publisher and Subscriber [#2394](https://github.com/eclipse-iceoryx/iceoryx/issues/2394)
 - Fix call to non-existing `getService` in channel.inl [#2426](https://github.com/eclipse-iceoryx/iceoryx/issues/2426)
 - Fix chunk management race condition causing failed management allocation [#2440](https://github.com/eclipse-iceoryx/iceoryx/issues/2440)
+- Fix compile error with GCC 8 [#2480](https://github.com/eclipse-iceoryx/iceoryx/issues/2480)
 
 **Refactoring:**
 

--- a/iceoryx_hoofs/reporting/include/iox/assertions.hpp
+++ b/iceoryx_hoofs/reporting/include/iox/assertions.hpp
@@ -80,7 +80,8 @@
 
 /// @brief panic if control flow reaches this code at runtime
 #define IOX_UNREACHABLE()                                                                                              \
-    iox::er::forwardPanic(IOX_CURRENT_SOURCE_LOCATION, "Reached code that was supposed to be unreachable.")
+    iox::er::detail::unreachable_wrapped(                                                                              \
+        true, true, IOX_CURRENT_SOURCE_LOCATION, "Reached code that was supposed to be unreachable.")
 
 // NOLINTEND(cppcoreguidelines-macro-usage)
 

--- a/iceoryx_hoofs/reporting/include/iox/assertions.hpp
+++ b/iceoryx_hoofs/reporting/include/iox/assertions.hpp
@@ -80,8 +80,8 @@
 
 /// @brief panic if control flow reaches this code at runtime
 #define IOX_UNREACHABLE()                                                                                              \
-    iox::er::detail::unreachable_wrapped(                                                                              \
-        true, true, IOX_CURRENT_SOURCE_LOCATION, "Reached code that was supposed to be unreachable.")
+    iox::er::detail::unreachable_wrapped<void, void>(IOX_CURRENT_SOURCE_LOCATION,                                      \
+                                                     "Reached code that was supposed to be unreachable.")
 
 // NOLINTEND(cppcoreguidelines-macro-usage)
 

--- a/iceoryx_hoofs/reporting/include/iox/error_reporting/error_forwarding.hpp
+++ b/iceoryx_hoofs/reporting/include/iox/error_reporting/error_forwarding.hpp
@@ -48,19 +48,13 @@ namespace detail
 {
 // workaround for gcc 8 bug
 // see also https://github.com/eclipse-iceoryx/iceoryx2/issues/855
-template <typename T, typename Message>
-[[noreturn]] constexpr inline void unreachable_wrapped(T t1, T t2, const SourceLocation& location, Message&& msg)
+template <typename T1, typename T2, typename Message>
+[[noreturn]] constexpr inline void unreachable_wrapped(const SourceLocation& location, Message&& msg)
 {
-#if (defined(__GNUC__) && (__GNUC__ >= 8 && __GNUC__ <= 8) && !defined(__clang__))
-    if (t1 == t2)
+    if (std::is_same<T1, T2>::value)
     {
         forwardPanic(location, std::forward<Message>(msg));
     }
-#else
-    static_cast<void>(t1);
-    static_cast<void>(t2);
-    forwardPanic(location, std::forward<Message>(msg));
-#endif
 }
 } // namespace detail
 

--- a/iceoryx_hoofs/reporting/include/iox/error_reporting/error_forwarding.hpp
+++ b/iceoryx_hoofs/reporting/include/iox/error_reporting/error_forwarding.hpp
@@ -44,6 +44,26 @@ template <typename Message>
     abort();
 }
 
+namespace detail
+{
+// workaround for gcc 8 bug
+// see also https://github.com/eclipse-iceoryx/iceoryx2/issues/855
+template <typename T, typename Message>
+[[noreturn]] constexpr inline void unreachable_wrapped(T t1, T t2, const SourceLocation& location, Message&& msg)
+{
+#if (defined(__GNUC__) && (__GNUC__ >= 8 && __GNUC__ <= 8) && !defined(__clang__))
+    if (t1 == t2)
+    {
+        forwardPanic(location, std::forward<Message>(msg));
+    }
+#else
+    static_cast<void>(t1);
+    static_cast<void>(t2);
+    forwardPanic(location, std::forward<Message>(msg));
+#endif
+}
+} // namespace detail
+
 /// @brief Forwards a fatal error and does not return.
 /// @param error the error
 /// @param kind the kind of error (category)

--- a/iceoryx_hoofs/reporting/include/iox/error_reporting/source_location.hpp
+++ b/iceoryx_hoofs/reporting/include/iox/error_reporting/source_location.hpp
@@ -23,7 +23,7 @@ namespace er
 {
 struct SourceLocation
 {
-    SourceLocation(const char* file, int line, const char* function)
+    constexpr SourceLocation(const char* file, int line, const char* function)
         : file(file)
         , line(line)
         , function(function)

